### PR TITLE
Universal ckp fixes

### PR DESCRIPTION
**Pull request by mosheisland**
This PR fixes few issues in loading from universal checkpoint.
With this fixes, it is ok to load from universal checkpoint and change all 3D dimensions (data / tensor / pipeline).
For testing of the changes, I used a 4-layers toy model of GPT-Large with BF16.

Following is TensorBoard with:
- Base run using DP=2 TP=2 PP=2 for 200 iterations
- Load from universal checkpoint at iteration 100 and run till iteration 200 with configurations:
  DP=1 TP=2 PP=2
  DP=2 TP=1 PP=2
  DP=1 TP=4 PP=2
  DP=1 TP=4 PP=1
  DP=8 TP=1 PP=1
  DP=4 TP=1 PP=2

As can be seen, both training loss and validation loss are similar across all runs.

universal-bf16-results

Note that this PR is dependent on https://github.com/microsoft/DeepSpeed/pull/4588

### CODE REVIEW

**Comment by tjruwase** [comment]
@stas00 FYI

File: megatron/model/gpt_model.py [hunk 1]
```diff
@@ -206,7 +206,7 @@ def universal_checkpoint_info(self):
             # Parameter that are sliced on the row dimension
             info[PARAMETER_WITH_ROW_PARALLELISM_PATTERNS] = [
                 r"\d+.mlp.dense_4h_to_h.weight",
-                r"\d+.mlp.self_attention.dense.weight",
+                r"\d+.self_attention.dense.weight",
```
Comment (stas00) [inline megatron/model/gpt_model.py:209]: this issue that you fixed is worrisome that it was there in the first place - which means that the checks aren't strict.

perhaps this mechanism needs to be redesigned to also validate that all rules have matched, and if they haven't - assert and flag to the user that some keys/regexes are wrong.

that is while going over the regex matches, a check is needed that there was at least 1 match.

File: megatron/checkpointing.py [hunk 2]
```diff
@@ -694,6 +693,19 @@ def load_checkpoint(model, optimizer, opt_param_scheduler, load_arg='load', stri
...
+            # the 'rng_tracker_states' is incorrect.
+            # to fix it, we reconfigure the seed of model parallel rng tracker.
+            # we use the iteration to ensure that we do not repeat the same random
+            # sequence for model parallel rng tracker.
+            ckp_args = state_dict['args']
+            if ((args.tensor_model_parallel_size != ckp_args.tensor_model_parallel_size)
+                    or (args.pipeline_model_parallel_size != ckp_args.pipeline_model_parallel_size)):
+                print_rank_0(' loading universal checkpoint with modified mp configuration '
+                             '-> reconfigure tp seed')
+                tensor_parallel.model_parallel_reconfigure_tp_seed(args.seed + iteration)
```
Comment (stas00) [inline megatron/checkpointing.py:707]: Thank you for flagging/attempting to fix this - this is indeed very important.

So basically this means that we don't really have a way to reshape a checkpoint and continue with the same stream where things left off and instead we are re-randomizing things and we no longer can ensure that there will be 1 epoch of unique data, correct?

And so the proposed solution at least removes duplicated streams, but this is still not what the trainer intended.

So at the moment it sounds like there is a need to have a disclaimer that if a checkpoint is reshaped then the unique stream of data for 1 epoch is no longer preserved.

But perhaps there is a correct technical solution this and not re-randomizing the streams.

Thoughts?

**Review by stas00 — COMMENTED** [review]
Thank you for the fixes, @mosheisland 

Please see my comments below

File: megatron/model/gpt_model.py (hunk 1 above)
Comment (mosheisland) [reply megatron/model/gpt_model.py:209]: Sounds reasonable. I will add it

File: megatron/checkpointing.py (hunk 2 above)
Comment (mosheisland) [reply megatron/checkpointing.py:707]: This change only reconfigures the **model-parallel-rng** RNG. The model-parallel-rng RNG affects 
(1) tensor parallel parameters initialization - not relevant in this case since we load from a checkpoint
(2) dropout ops within tensor parallel regions
But, it does not affect the data stream.

Comment (stas00) [reply megatron/checkpointing.py:707]: I was wondering about that - I guess I misread your comment's "the same random sequence" to mean data, but you meant RNG sequence.

So now that you have clarified that part, could you please clarify in that comment why `rng_tracker_states' is incorrect. 

Did you mean to say it shouldn't be the same if we are adding more resources, to improve the learning? After all the data will be different so what would be the actual issue if we 2x TP and the rng will be the same?

Comment (stas00) [reply megatron/checkpointing.py:707]: I haven't touched this code in a long long time, so it's hard for me to remember all the nuances of this super complex machinery. So as your view is fresh please bear with me.

Comment (mosheisland) [reply megatron/checkpointing.py:707]: The reason why we have model-parallel-rng is that each tp-worker performs dropout on attention_probs tp slice.
In order to prevent same dropout mask in the different tp slices, we assign a unique RNG, model-parallel-rng, for each TP worker and use it using the fork mechanism. Similar explanation applies to why we (optionally, when we don't initialize on cpu) use model-parallel-rng when we initialize TP-split parameters (e.g. ColumnParallelLinear, RowParallelLinear).

When we load from universal checkpoint, we do have the mp_rank_<mp>_model_states.pt checkpoint files that contain the model-parallel-rng per each {tp-rank, pp-rank} combination. However, think of the following scenario:
Assuming we are loading from universal checkpoint that was using DP=1, PP=1, TP=2 to current configuration of DP=1, PP=1, TP=4. In this case, we don't have model-parallel-rng information for tp ranks 3 and 4. Also, when we move from TP=4 to TP=2, we can use 2 of the 4 rng states, but this will not provide any gain in terms of reproducibility.
And, in any case, the dropout random patterns will be different since we will be using 4 different RNGs (tp=4) instead of 2 up till now.
Therefore, I don't see a simple way that we can maintain dropout pattern reproducibility going forward when we modify TP.

During initialization, each tp rank is initialized with its own unique tp_seed derived from args.seed (_tp_seed = args.seed + tp_rank +constant_). I could keep it as-is, but I was trying to avoid the case where we load multiple times from universal, every time with different TP, during the course of training. This would have led to the same dropout patterns being used repeatedly. I am avoiding this by reconfiguring model-parallel-rng seed to args.seed + **iteration** + tp_rank + constant.

Hope this clarifies.

Comment (stas00) [reply megatron/checkpointing.py:707]: Wow, thank you for the amazing explanation, Moshe - that's very helpful.

How about we dump your explanation verbatim somewhere in the vicinity of the new code? 

with a summary:
TLDR: unique rng is needed for dropout to be really random on TP ranks

Comment (mosheisland) [reply megatron/checkpointing.py:707]: Sure, will do.

**Comment by mosheisland** [comment]
@tjruwase, I think that now that https://github.com/microsoft/DeepSpeed/pull/4588 has been merged, this PR can also be merged.
